service/s3/endpoints.go

```diff
@@ -213,6 +215,127 @@ func finalizeClientEndpointResolverOptions(options *Options) {
+	legacyResolver EndpointResolver
+	resolver       EndpointResolverV2
+}
+
+func (l *legacyEndpointResolverAdapter) ResolveEndpoint(ctx context.Context, params EndpointParameters) (endpoint smithyendpoints.Endpoint, err error) {
+	var fips aws.FIPSEndpointState
+	var dualStack aws.DualStackEndpointState
+
+	if aws.ToBool(params.UseFIPS) {
+		fips = aws.FIPSEndpointStateEnabled
+	}
+	if aws.ToBool(params.UseDualStack) {
+		dualStack = aws.DualStackEndpointStateEnabled
+	}
+
+	resolveEndpoint, err := l.legacyResolver.ResolveEndpoint(aws.ToString(params.Region), EndpointResolverOptions{
+		ResolvedRegion:       aws.ToString(params.Region),
+		UseFIPSEndpoint:      fips,
+		UseDualStackEndpoint: dualStack,
+	})
+	if err != nil {
+		return endpoint, err
+	}
+
+	if resolveEndpoint.HostnameImmutable {
+		uriString := resolveEndpoint.URL
+		uri, err := url.Parse(uriString)
+		if err != nil {
+			return endpoint, fmt.Errorf("Failed to parse uri: %s", uriString)
+		}
+
+		return smithyendpoints.Endpoint{
+			URI: *uri,
+		}, nil
+	}
+
+	if resolveEndpoint.Source == aws.EndpointSourceServiceMetadata {
+		return l.resolver.ResolveEndpoint(ctx, params)
+	}
+
+	params.UseDualStack = nil
+	params.UseFIPS = nil
+	params.Endpoint = &resolveEndpoint.URL
+
+	return l.resolver.ResolveEndpoint(ctx, params)
+}
+
+type isDefaultProvidedImplementation interface {
+	isDefaultProvidedImplementation()
+}
+
+type compatibleEndpointResolver struct {
+	EndpointResolverV2 EndpointResolverV2
+}
+
+func (n *compatibleEndpointResolver) isDefaultProvidedImplementation() {}
+
+func (n *compatibleEndpointResolver) ResolveEndpoint(region string, options EndpointResolverOptions) (endpoint aws.Endpoint, err error) {
+	reg := region
+	fips := options.UseFIPSEndpoint
+	if len(options.ResolvedRegion) > 0 {
+		reg = options.ResolvedRegion
+	} else {
+		reg, fips = mapPseudoRegion(region)
+	}
+	ctx := context.Background()
+	resolved, err := n.EndpointResolverV2.ResolveEndpoint(ctx, EndpointParameters{
+		Region:       &reg,
+		UseFIPS:      aws.Bool(fips == aws.FIPSEndpointStateEnabled),
+		UseDualStack: aws.Bool(options.UseDualStackEndpoint == aws.DualStackEndpointStateEnabled),
+	})
+	if err != nil {
+		return endpoint, err
+	}
+
+	endpoint = aws.Endpoint{
+		URL:               resolved.URI.String(),
+		HostnameImmutable: false,
+		Source:            aws.EndpointSourceServiceMetadata,
+	}
+
+	return endpoint, nil
+}
+
+func mapPseudoRegion(pr string) (region string, fips aws.FIPSEndpointState) {
+	const fipsInfix = "-fips-"
+	const fipsPrefix = "fips-"
+	const fipsSuffix = "-fips"
+
+	if strings.Contains(pr, fipsInfix) ||
+		strings.Contains(pr, fipsPrefix) ||
+		strings.Contains(pr, fipsSuffix) {
+		region = strings.ReplaceAll(strings.ReplaceAll(strings.ReplaceAll(
+			pr, fipsInfix, "-"), fipsPrefix, ""), fipsSuffix, "")
+		fips = aws.FIPSEndpointStateEnabled
+	} else {
+		region = pr
+	}
+
+	return region, fips
+}
+
+func finalizeEndpointResolverV2(options *Options) {
```

Comment:
Is this needed for something down the road? I'm not seeing it used anywhere, only generated.

Reply:
yah this will be needed later down the road. ill add it to this PR